Add Lazy Ingestion Mechanism for PyPI

gdrosos opened the pull request:
## Description
This PR aims to extend REST-API plugin to support a lazy ingestion mechanism for PyPI:

More precisely, we extended the methods `ingestArtifactIfNecessary()` and `ingestArtifactWithDependencies()` in order to ingest single pypi coordinates and the transitive closure of their dependencies.

* In the `ingestArtifactIfNecessary()` method we search if the specific coordinate exists in the db, and if not we check through the PyPI Rest-API if it exists on PyPI. If this is a case, we produce in an output topic the  information of the specific coordinate retrieved from the rest-api, which is then  processed and checked for deduplication by the [PyPI Filter ](https://github.com/fasten-project/pypi-tools/pull/4)tool and consequently fed to the PyPI pipeline.
* In the `ingestArtifactWithDependencies()` method we employ the pypi-resolver to retrieve the transitive dependencies of the coordinate and for each coordinate we call the `ingestArtifactIfNecessary()` method. 
* We enabled triggering  the `ingestArtifactWithDependencies()` method when the dependency resolution endpoint is triggered for a specific coordinate
* We also increased the maximum request size of the kafka producer to 3mb, in order to be able to handle large records produxed by the PyPI Api, like the on of the [cryptography](https://pypi.org/project/cryptography/) cooridnate, which previously could not be processed due to limited request size.

## Motivation and context
The Fasten Python pipeline did not support a lazy-ingestion mechanism, making it difficult to use the REST-API for various analysies. An ingestion mechanism will allow users to observe fine-grained information of a specific requested PyPI coordinate faster.

## Testing
Tested the changes locally and on docker compose, on the [pypi-lazy-ingestion branch](https://github.com/fasten-project/fasten-docker-deployment/tree/pypi-lazy-ingestion). The cooridnates could be succesfully be fed to the python pipeline, along with the transitive closure of their dependencies
## Task list  
- [x] Add lazy-ingestion for a single coordinate
- [x] Add lazy-ingestion for the transitive closure of the specific coordinate
- [x] Test the REST-API on docker compose
- [x] Resolved conflicts via rebasing

MagielBruntink reviewed (changes requested):
Great stuff @gdrosos this functionality is really needed.

Some comments:

- The mvn lazy ingest request checks whether the coordinate has not already been requested for ingestion before: `if (!hasArtifactBeenIngested(packageName, version)) {` It seems the PyPI case does not do this. How does it prevent requesting the same package-version many times, in particular when doing transitive closure requests?

- I'm not a fan of the switch/cases added in the existing methods with full implementation inside. Could the ecosystem-specific parts be pulled out into their specific helper methods? So that the switch/case only calls the right method, and the entire method size remains low.

gdrosos commented:
> Great stuff @gdrosos this functionality is really needed.
> 
> Some comments:
> 


Thanks @MagielBruntink,
> * The mvn lazy ingest request checks whether the coordinate has not already been requested for ingestion before: `if (!hasArtifactBeenIngested(packageName, version)) {` It seems the PyPI case does not do this. How does it prevent requesting the same package-version many times, in particular when doing transitive closure requests?

With the current implementation, we will produce as many records to the output topic as the number of unknown cooridnates requested/retrieved though the transitive closure. But then the PyPI Filter keeps track through a list of all the processed coordinates and will filter the duplicates and feed to the python pipeline only the ones which have not been ingested.

> * I'm not a fan of the switch/cases added in the existing methods with full implementation inside. Could the ecosystem-specific parts be pulled out into their specific helper methods? So that the switch/case only calls the right method, and the entire method size remains low.

So what you are proposing is separating the concerns in the ingestion methods so that each language specific process is separated into a specific method? I think that this refactoring is feasible.

mir-am commented:
Good job, @gdrosos. 
I will review the code tomorrow morning.

mir-am commented on analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/KnowledgeBaseConnector.java:
```diff
@@ -210,14 +210,14 @@ public void connectToGraphDB() {
             producerProperties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
             producerProperties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
-            producerProperties.setProperty(ProducerConfig.MAX_REQUEST_SIZE_CONFIG, "1000000");
+            producerProperties.setProperty(ProducerConfig.MAX_REQUEST_SIZE_CONFIG, "3000000");
```
Why is `MAX_REQUEST_SIZE_CONFIG` increased? for sending larger records?

mir-am commented on analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/LazyIngestionProvider.java:
```diff
@@ -18,9 +18,14 @@
 
 package eu.fasten.analyzer.restapiplugin;
 
+
```
Should this extra blank line be removed here?

gdrosos replied:
Remvoed on [28d46c8](https://github.com/fasten-project/fasten/pull/470/commits/28d46c85d7de232b6ede4bd34aa57fdaaa3e02da)

mir-am commented on analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/LazyIngestionProvider.java:
```diff
@@ -43,47 +48,93 @@ private static String toKey(String packageName, String version) {
-            if (date != null && date > 0) {
-            	// TODO why is the date necessary?
-                jsonRecord.put("date", date);
+        switch(KnowledgeBaseConnector.forge){
```
As Magiel pointed out, please create helper methods for the logic inside the switch cases to reduce the complexity of the `ingestArtifactIfNecessary` method.

gdrosos replied:
Fixed on [28d46c8](https://github.com/fasten-project/fasten/pull/470/commits/28d46c85d7de232b6ede4bd34aa57fdaaa3e02da)

mir-am reviewed (commented):
Overall, nice work, @gdrosos. Aside from my comments in the review, the PyPI's lazy ingestion mechanism should avoid duplicates, i.e., package versions that are already ingested into the pipeline. Perhaps, you can adopt a similar approach to Maven, creating an ingestion table for checking. This should also work when scaling your application horizontally.

gdrosos commented:
Thanks for all the insightful comments! I tried to separate the concerns in the ingestion methods on 28d46c8.
Regarding the deduplication mechanism , the problem is that on the python db schema the ingested_artifacts table has the previous schema which has the packageName, version and timestamp attributes. The problem is that on the MetadataDAO of the Fasten Server the new schema is used,meaning that if I try to retrieve the `IngestedArtifacts.INGESTED_ARTIFACTS.VERSION` constant (in the source code) an error will yield  as only the following attributes exist:
```
IngestedArtifacts.INGESTED_ARTIFACTS.KEY
IngestedArtifacts.INGESTED_ARTIFACTS.PLUGIN_VERSION
IngestedArtifacts.INGESTED_ARTIFACTS.TIMESTAMP
```
. If I attempt to use those Constants also an error will yield when deploying the rest-api, because on the python db those attributes do not exist. So how can I tackle this issue?

MagielBruntink commented:
> Thanks for all the insightful comments! I tried to separate the concerns in the ingestion methods on [28d46c8](https://github.com/fasten-project/fasten/commit/28d46c85d7de232b6ede4bd34aa57fdaaa3e02da). Regarding the deduplication mechanism , the problem is that on the python db schema the ingested_artifacts table has the previous schema which has the packageName, version and timestamp attributes. The problem is that on the MetadataDAO of the Fasten Server the new schema is used,meaning that if I try to retrieve the `IngestedArtifacts.INGESTED_ARTIFACTS.VERSION` constant (in the source code) an error will yield as only the following attributes exist:
> 
> ```
> IngestedArtifacts.INGESTED_ARTIFACTS.KEY
> IngestedArtifacts.INGESTED_ARTIFACTS.PLUGIN_VERSION
> IngestedArtifacts.INGESTED_ARTIFACTS.TIMESTAMP
> ```
> 
> . If I attempt to use those Constants also an error will yield when deploying the rest-api, because on the python db those attributes do not exist. So how can I tackle this issue?

Hi Georgios,

This is an oversight. The `ingested_artifacts` table needs to be re-created on the Python (and C) DBs. @mir-am can take care of this on monster. 
```
DROP TABLE ingested_artifacts;

CREATE TABLE ingested_artifacts
(
    -- leave the key-contents to the plugin implementation
    key            TEXT PRIMARY KEY,
    plugin_version TEXT NOT NULL,
    timestamp      TIMESTAMP
);
```
I have updated the docker-compose DB schemas to match. That should take care of the issue.

https://github.com/fasten-project/fasten-docker-deployment/commit/ed1e387788ab09b2bf3f3d0bd166d96b24741bc4

mir-am commented:
I have added the `ingested_artifacts` with the above-specified fields to the production DB and to the SQL scripts.

gdrosos commented:
Hello @MagielBruntink, @mir-am

Thanks for all the insightful comments, I tried to address them.

 More specifically, I simplified the ingestion methods by separating the body of the switch cases into separate, language specific helper methods.

 I also added the proposed deduplication mechanism: Before ingesting the specific coordinate we check whether this coordinate has been ingested (the key of the package:version exists in the `ingested_artifacts` table) and if this is not the case we proceed with producing an output record to kafka.

mir-am reviewed (approved):
@gdrosos, thanks for addressing the comments. 
The changes LGTM!

gdrosos commented:
I tested the changes on DC, and while the plugin seemed to work fine, the records could not be succesfully inserted into the db due to an issue in the MetadataPlugin. Strangely, I could encounter this issue only after populating the ingestion table, as before this commit (as far as I can remember) the ingested coordinates could be populated in the db. 
```
{
  "createdAt": 1654693559923,
  "input": {
    "input": {
      "requires_dist": [],
      "product": "pycg-stitch",
      "version_timestamp": 1652184206,
      "version": "0.0.1"
    },
    "plugin_version": "0.0.1",
    "payload": {
      "product": "pycg-stitch",
      "metadata": {
        "loc": 332,
        "max_rss": 20056,
        "time_elapsed": 0.18,
        "num_files": 6
      },
      "nodes": 45,
      "forge": "PyPI",
      "generator": "PyCG",
      "depset": [],
      "version": "0.0.1",
      "modules": {
        "internal": {
          ...
      },
      "sourcePath": "/mnt/fasten/revision-callgraphs/pypi/pypi/sources/p/pycg-stitch/0.0.1",
      "timestamp": "1652184206"
    },
    "created_at": 1654693559,
    "plugin_name": "PyCG"
  },
  "plugin": "eu.fasten.analyzer.metadataplugin.MetadataDatabasePythonPlugin$MetadataDBPythonExtension",
  "host": "1c7b6ba315ea",
  "consumedAt": 1654693559915,
  "error": {
    "stacktrace": "org.jooq.exception.DataAccessException: SQL [insert into \"public\".\"packages\" (\"package_name\", \"forge\") values (?, ?) on conflict on constraint \"unique_package_forge\" do update set \"package_name\" = \"excluded\".\"package_name\", \"forge\" = \"excluded\".\"forge\" returning \"public\".\"packages\".\"id\"]; ERROR: constraint \"unique_package_forge\" for table \"packages\" does not exist\n\tat org.jooq_3.16.3.POSTGRES.debug(Unknown Source)\n\tat org.jooq.impl.Tools.translate(Tools.java:3089)\n\tat org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:670)\n\tat org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:354)\n\tat org.jooq.impl.DMLQueryAsResultQuery.fetch(DMLQueryAsResultQuery.java:143)\n\tat org.jooq.impl.ResultQueryTrait.fetchLazy(ResultQueryTrait.java:278)\n\tat org.jooq.impl.ResultQueryTrait.fetchLazyNonAutoClosing(ResultQueryTrait.java:287)\n\tat org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:506)\n\tat eu.fasten.core.data.metadatadb.MetadataDao.insertPackage(MetadataDao.java:146)\n\tat eu.fasten.analyzer.metadataplugin.MetadataDBExtension.saveToDatabase(MetadataDBExtension.java:230)\n\tat eu.fasten.analyzer.metadataplugin.MetadataDBExtension.lambda$consume$0(MetadataDBExtension.java:138)\n\tat org.jooq.impl.DefaultDSLContext.lambda$transaction$5(DefaultDSLContext.java:612)\n\tat org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$3(DefaultDSLContext.java:550)\n\tat org.jooq.impl.Tools$3$1.block(Tools.java:5600)\n\tat java.base/java.util.concurrent.ForkJoinPool.managedBlock(Unknown Source)\n\tat org.jooq.impl.Tools$3.get(Tools.java:5597)\n\tat org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:598)\n\tat org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:522)\n\tat org.jooq.impl.DefaultDSLContext.transaction(DefaultDSLContext.java:611)\n\tat eu.fasten.analyzer.metadataplugin.MetadataDBExtension.consume(MetadataDBExtension.java:133)\n\tat eu.fasten.core.plugins.KafkaPlugin.consume(KafkaPlugin.java:59)\n\tat eu.fasten.server.plugins.kafka.FastenKafkaPlugin.processRecord(FastenKafkaPlugin.java:340)\n\tat eu.fasten.server.plugins.kafka.FastenKafkaPlugin.handleConsuming(FastenKafkaPlugin.java:270)\n\tat eu.fasten.server.plugins.kafka.FastenKafkaPlugin.run(FastenKafkaPlugin.java:167)\n\tat eu.fasten.server.plugins.kafka.FastenKafkaPlugin.start(FastenKafkaPlugin.java:205)\n\tat java.base/java.util.ArrayList.forEach(Unknown Source)\n\tat eu.fasten.server.FastenServer.run(FastenServer.java:202)\n\tat picocli.CommandLine.executeUserObject(CommandLine.java:1919)\n\tat picocli.CommandLine.access$1100(CommandLine.java:145)\n\tat picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)\n\tat picocli.CommandLine$RunLast.handle(CommandLine.java:2326)\n\tat picocli.CommandLine$RunLast.handle(CommandLine.java:2291)\n\tat picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)\n\tat picocli.CommandLine.execute(CommandLine.java:2058)\n\tat eu.fasten.server.FastenServer.main(FastenServer.java:435)\nCaused by: org.postgresql.util.PSQLException: ERROR: constraint \"unique_package_forge\" for table \"packages\" does not exist\n\tat org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)\n\tat org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)\n\tat org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)\n\tat org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)\n\tat org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)\n\tat org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)\n\tat org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:118)\n\tat org.jooq.tools.jdbc.DefaultPreparedStatement.executeQuery(DefaultPreparedStatement.java:104)\n\tat org.jooq.impl.AbstractDMLQuery.executeReturningQuery(AbstractDMLQuery.java:1249)\n\tat org.jooq.impl.AbstractDMLQuery.execute(AbstractDMLQuery.java:1074)\n\tat org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:340)\n\t... 31 more\n",
    "type": "org.jooq.exception.DataAccessException",
    "message": "SQL [insert into \"public\".\"packages\" (\"package_name\", \"forge\") values (?, ?) on conflict on constraint \"unique_package_forge\" do update set \"package_name\" = \"excluded\".\"package_name\", \"forge\" = \"excluded\".\"forge\" returning \"public\".\"packages\".\"id\"]; ERROR: constraint \"unique_package_forge\" for table \"packages\" does not exist"
  },
  "version": "0.1.2"
}
```

MagielBruntink commented:
Strange, an index is missing, on the packages table:
ERROR: constraint \"unique_package_forge\" for table \"packages\" does not exist

That constraint index should be added be the DB creation script… could you see in the metadata db container log if the db creation was successful?

gdrosos commented:
> Strange, an index is missing, on the packages table: ERROR: constraint "unique_package_forge" for table "packages" does not exist
> 
> That constraint index should be added be the DB creation script… could you see in the metadata db container log if the db creation was successful?

Yes, the db creation was succesfull.
I am attatching the logs from the db container:
```
metadata-db_1                                       | 2022-06-08 17:01:24.736 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
metadata-db_1                                       | 2022-06-08 17:01:24.736 UTC [1] LOG:  listening on IPv6 address "::", port 5432
metadata-db_1                                       | 2022-06-08 17:01:25.016 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
metadata-db_1                                       | 2022-06-08 17:01:25.423 UTC [26] LOG:  database system was shut down at 2022-06-08 17:01:02 UTC
metadata-db_1                                       | 2022-06-08 17:01:25.580 UTC [1] LOG:  database system is ready to accept connections
metadata-db_1                                       | 2022-06-08 17:03:32.313 UTC [35] ERROR:  constraint "unique_package_forge" for table "packages" does not exist
metadata-db_1                                       | 2022-06-08 17:03:32.313 UTC [35] STATEMENT:  insert into "public"."packages" ("package_name", "forge") values ($1, $2) on conflict on constraint "unique_package_forge" do update set "package_name" = "excluded"."package_name", "forge" = "excluded"."forge" returning "public"."packages"."id
```

I tried also feeding normally the pipeline with the example coordinates on DC and I received the error also for them.

MagielBruntink commented:
My bad, those indices should indeed have been removed on the C and Python DBs as well. 
@mir-am could you please remove the `unique_ingested_artifacts` indices as well?

MagielBruntink commented:
I have updated DC, that should take care of it: https://github.com/fasten-project/fasten-docker-deployment/commit/66d48d9dc40f8839ceafe55f79b80c73482ceea1

mir-am commented:
> My bad, those indices should indeed have been removed on the C and Python DBs as well. @mir-am could you please remove the `unique_ingested_artifacts` indices as well?
@MagielBruntink 
It seems that the `unique_ingested_artifacts` index does not exist for both C and Python DBs in production.

MagielBruntink commented:
Cool, makes sense @mir-am

gdrosos commented:
The db issue was created due to the existence of the  `unique_ingested_artifacts` indice which did not enable the correct db initialization. 
I tested on DC the changes with the correct schema, and everything works fine (the records are successfully processed and the deduplication mechanism works).
Therefore, I am rebasing this branch into `develop`

In order to apply those changes, we need to publish a new rest-api release.
